targets/TARGET_Samsung/TARGET_SIDK_S5JS100/sleep.c

```diff
@@ -0,0 +1,160 @@
+/****************************************************************************
+ *
+ * Copyright 2020 Samsung Electronics All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied. See the License for the specific
+ * language governing permissions and limitations under the License.
+ *
+ ****************************************************************************/
+#if DEVICE_SLEEP
+
+#include <stdio.h>
+#include <string.h>
+#include "sleep_api.h"
+#include "cmsis.h"
+#include "gpio_api.h"
+#include "mbed_power_mgmt.h"
+#include "mbed_critical.h"
+#include "rtos_idle.h"
+
+int nbsleep_req = 0;
+int external_pin = 0;
+static int initialize_policy = 0;
+static int disable_poweroffbycp = 1;
+
+static inline void hw_delay_us(unsigned int value)
+{
+    volatile unsigned i, j;
+    for (i = 0; i < (value * 2); i++) {
+        for (j = 0; j < 100; j++);
+    }
+}
+
+int s5js100_idle_sicd(void)
+{
+    if (getreg32(0x85023090) & (0x1 << 1)) {
+        /* send CP to ready status */
+        modifyreg32(0x85023090, (0x1 << 2), (0x1 << 2));
+        modifyreg32(0x85026100, 0x2, 0x0);
+
+        pmu_short_sleep(); //go to sleep
+
+        modifyreg32(0x85026100, 0x2, 0x2);
+
+        /* wait for mbox */
+        while ((getreg32(0x85023090) & (0x1 << 1))) {
+            hw_delay_us(10);
+            putreg32(0x2, 0x85023020);
+            putreg32(0x2, 0x8502301C);
+        };
+
+        /* release short sleep status */
+        modifyreg32(0x85023090, (0x1 << 2), (0x0 << 2));
+        return 1;
+    } else {
+        asm("wfi");
+    }
+
+    return 0;
+}
+
+static void s5js100_idle_hook(void)
+{
+    core_util_critical_section_enter();
+    sleep();
+    core_util_critical_section_exit();
+}
+
+static void set_sleep_policy(void)
+{
+    rtos_attach_idle_hook(&s5js100_idle_hook);
```

Comment:
Sorry to raise this since the PR is merged already, but I'm wondering if we can drop this idle hook and use the default one?

The Mbed OS default idle hook supports Tickless:
https://github.com/ARMmbed/mbed-os/blob/b1629b7e5962f9fd0382e6b9bb83e97b6f9f6b13/rtos/source/TARGET_CORTEX/mbed_rtx_idle.cpp#L133-L154

Also, calling `rtos_attach_idle_hook` here means the `hal_sleep()` implementation can't compile with the bare metal profile.

@andrewc-arm @kjbracey-arm

Reply:
@vince-zeng can you take a look together with @kjbracey-arm  ?
@LDong-Arm , FYI andrewc-arm no longer works for Arm.

Reply:
I have forward the info to Samsung team and please them help to check. Thanks!

Reply:
In default_idle_hook(), the only different with s5js100_idle_hook() is that the sleep_manager_lock_deep_sleep_internal() is called. 
But I am not familiar with this function, could you please share about this?

Reply:
@LDong-Arm Can you create an issue with details and we can discuss it there and fix?

Reply:
#13265 created